torch/csrc/autograd/variable.cpp

```diff
@@ -106,52 +144,84 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
   attr_version = version_counter.current_version();
 }
 
-std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+std::shared_ptr<Function>& Variable::ViewImpl::get_grad_fn() {
   std::lock_guard<std::mutex> lock(mutex);
-  if (!_grad_fn && !base.requires_grad()) {
-    return _grad_fn;
+  if (!grad_fn && !base.requires_grad()) {
+    return grad_fn;
   }
   auto current_version = version_counter.current_version();
   if (attr_version != current_version) {
     TORCH_ASSERT(output_nr == 0);
     auto fn = std::make_shared<generated::AsStridedBackward>();
-    fn->self_geometry = TensorGeometry(base);
+    fn->self_geometry = at::TensorGeometry(base);
     fn->size = sizes();
     fn->stride = strides();
     fn->storage_offset = data.storage_offset();
     fn->set_next_functions(get_next_functions(base));
     fn->num_inputs = 1;
-    _grad_fn = std::move(fn);
+    grad_fn = std::move(fn);
     attr_version = current_version;
   }
-  return _grad_fn;
+  return grad_fn;
 }
 
-void VariableViewImpl::rebase_history(int output_nr, std::shared_ptr<Function> grad_fn) {
+void Variable::ViewImpl::rebase_history(
+    int output_nr,
+    std::shared_ptr<Function> grad_fn) {
   TORCH_ASSERT(output_nr == 0);
   TORCH_ASSERT(grad_fn);
-  TORCH_ASSERTM(grad_fn->num_inputs == 1, "Functions which modify views in-place must return a single Variable");
+  TORCH_ASSERTM(
+      grad_fn->num_inputs == 1,
+      "Functions which modify views in-place must return a single Variable");
   this->output_nr = output_nr;
-  base.output_nr() = 0;
-  base.get()->_grad_fn = std::make_shared<CopySlices>(
-      base, TensorGeometry(data), std::move(grad_fn));
-  get_grad_fn();  // trigger an update to the view's grad_fn
+  auto copy_slices = std::make_shared<CopySlices>(
+      base, at::TensorGeometry(data), std::move(grad_fn));
+  base.set_gradient_edge({std::move(copy_slices), 0});
+  get_grad_fn(); // trigger an update to the view's grad_fn
+}
+
+Variable::Impl* Variable::get() const {
+  return static_cast<Variable::Impl*>(pImpl);
+}
+
+void Variable::rebase_history(
+    int output_nr,
+    std::shared_ptr<Function> grad_fn) {
+  TORCH_ASSERT(grad_fn);
+  if (is_view()) {
+    auto& impl = static_cast<Variable::ViewImpl&>(*get());
+    impl.rebase_history(output_nr, std::move(grad_fn));
+  } else {
+    get()->output_nr = output_nr;
+    get()->grad_fn = std::move(grad_fn);
+  }
+}
+
+bool Variable::requires_grad() const noexcept {
+  return get()->requires_grad || get()->grad_fn ||
+      (is_view() && base().requires_grad());
```

Comment:
Why is this no longer in the header? It's used all over the place, and it should be inlined.

Reply:
My bad, will put back.